File: app/src/main/java/com/example/himasha/workhub/SingleJobView.java
package com.example.himasha.workhub;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

public class SingleJobView extends AppCompatActivity {

    private String job_key = null;
    private String jobAddedUser = null;

    private TextView jobName;
    private TextView jobDesc;
    private TextView jobLocation;
    private TextView jobBudget;
    private TextView jobPostedDate;
    private TextView jobPostedUser;
    private TextView jobPostedMap;

    private Button deletejob;
    private Button editjob;
    private Button saveJob;
    private Button feedbackJob;

    private RecyclerView sugges_list;

    private CardView locationMapCardview;
    private CardView editRemoveJobcardview;
    private CardView postedUserCardview;

    private DatabaseReference workhubUsers;
    private DatabaseReference workhubJobs;
    private DatabaseReference workhubSavedJobs;
    private DatabaseReference workhubsuges;
    private FirebaseAuth auth;

    private AlertDialog.Builder builder;
    private String job_location;

    private Double jobLat;
    private Double jobLong;
    private String posteduserid;

    private String keyword;

    private SharedPreferences pref;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_single_job_view);
        pref = this.getSharedPreferences("Users",0);
        job_key = getIntent().getExtras().getString("job_id");
        auth = FirebaseAuth.getInstance();
        workhubJobs = FirebaseDatabase.getInstance().getReference().child("jobs");
        workhubSavedJobs = FirebaseDatabase.getInstance().getReference().child("saved_jobs");
        workhubUsers = FirebaseDatabase.getInstance().getReference().child("users");

        builder = new AlertDialog.Builder(this);

        jobName = (TextView) findViewById(R.id.singlejobName);
        jobDesc = (TextView) findViewById(R.id.singlejobDesc);
        jobLocation = (TextView) findViewById(R.id.singlejobLocation);
        jobBudget = (TextView) findViewById(R.id.singlejobBudget);
        jobPostedDate = (TextView) findViewById(R.id.singlejobPostedDate);
        jobPostedUser = (TextView) findViewById(R.id.singlejobPostedUser);
        jobPostedMap = (TextView) findViewById(R.id.singlejobMapName);

        deletejob = (Button) findViewById(R.id.singlejobDeleteBTN);
        saveJob = (Button) findViewById(R.id.saveJobButton);
        editjob = (Button) findViewById(R.id.singlejobEditBTN);
        feedbackJob = (Button) findViewById(R.id.feedbackButton);

        sugges_list = (RecyclerView)findViewById(R.id.sugges_list);
        sugges_list.setHasFixedSize(true);
        sugges_list.setLayoutManager(new LinearLayoutManager(this));

        locationMapCardview = (CardView)findViewById(R.id.locationMapCard);
        editRemoveJobcardview = (CardView)findViewById(R.id.editremoveJobCard);
        postedUserCardview = (CardView) findViewById(R.id.postedUserCard);


        workhubJobs.child(job_key).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot1) {

                keyword = (String) dataSnapshot1.child("jobKeyWord").getValue();
                final String job_name = (String) dataSnapshot1.child("jobName").getValue();
                final String job_desc = (String) dataSnapshot1.child("jobDesc").getValue();
                job_location = (String) dataSnapshot1.child("jobLocationName").getValue();
                final String job_budget = (String) dataSnapshot1.child("jobBudget").getValue();
                final String job_postedDate = (String) dataSnapshot1.child("jobPostedDate").getValue();
                final String job_postedUser = (String) dataSnapshot1.child("jobPostedUserId").getValue();
                posteduserid = (String) dataSnapshot1.child("jobPostedUserId").getValue();

                final String job_postedUserName = (String) dataSnapshot1.child("jobPostedUserName").getValue();
                jobLat = (Double) dataSnapshot1.child("jobLocationLat").getValue();
                jobLong = (Double) dataSnapshot1.child("jobLocationLong").getValue();

                jobName.setText(job_name);
                jobDesc.setText(job_desc);
                jobLocation.setText(job_location);
                jobBudget.setText("Rs."+job_budget);
                jobPostedDate.setText("Posted on "+job_postedDate);
                jobPostedUser.setText("Posted by "+job_postedUserName);

                jobPostedMap.setText("get directions to "+job_location);


                if (auth.getCurrentUser().getUid().equals(job_postedUser))
                {
                    editRemoveJobcardview.setVisibility(View.VISIBLE);
                }
                else {
                    editRemoveJobcardview.setVisibility(View.GONE);
                }

            }
            @Override
            public void onCancelled(DatabaseError databaseError) {

            }

        });

        postedUserCardview.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (posteduserid.equals(auth.getCurrentUser().getUid()))
                {
                    Intent intent = new Intent(SingleJobView.this, SingleProfileActivity.class);
                    intent.putExtra("user_id", posteduserid);
                    startActivity(intent);
                }
                else {
                    Intent intent = new Intent(SingleJobView.this, SingleProfileActivity.class);
                    intent.putExtra("user_id", posteduserid);
                    startActivity(intent);
                }

            }
        });

        locationMapCardview.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {


                Intent intent = new Intent(SingleJobView.this, MapLoadActivity.class);
                intent.putExtra("job_lat", jobLat);
                intent.putExtra("job_long",jobLong);
                startActivity(intent);
            }
        });


        editjob.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(SingleJobView.this, EditJobActivity.class);
                intent.putExtra("job_id", job_key);
                startActivity(intent);
            }
        });

        deletejob.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                builder.setTitle("Confirm");
                builder.setMessage("Are you sure you want to delete this job?");

                builder.setPositiveButton("YES", new DialogInterface.OnClickListener() {

                    public void onClick(DialogInterface dialog, int which) {
                        if(job_key != null) {
                            workhubJobs.child(job_key).removeValue();

                            Intent intent2 = new Intent(SingleJobView.this, FeedActivity.class);
                            startActivity(intent2);
                            finish();
                        }

                        dialog.dismiss();
                    }
                });

                builder.setNegativeButton("NO", new DialogInterface.OnClickListener() {

                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                        // Do nothing
                        dialog.dismiss();
                    }
                });

                AlertDialog alert = builder.create();
                alert.show();

            }
        });

        saveJob.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                builder.setTitle("Confirm");
                builder.setMessage("Are you sure you want to save this job?");

                builder.setPositiveButton("YES", new DialogInterface.OnClickListener() {

                    public void onClick(DialogInterface dialog, int which) {
                        workhubJobs.child(job_key).addValueEventListener(new ValueEventListener() {
                            @Override
                            public void onDataChange(DataSnapshot dataSnapshot) {
                                dataSnapshot.getRef().child("isSaved").setValue("1");
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {

                            }
                        });

                        dialog.dismiss();
                    }
                });

                builder.setNegativeButton("NO", new DialogInterface.OnClickListener() {

                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                        // Do nothing
                        dialog.dismiss();
                    }
                });

                AlertDialog alert = builder.create();
                alert.show();

            }
        });


        feedbackJob.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                builder.setTitle("Confirm");
                builder.setMessage("Are you sure you want to save this job?");

                builder.setPositiveButton("YES", new DialogInterface.OnClickListener() {

                    public void onClick(DialogInterface dialog, int which) {
                        if(job_key != null) {
                            workhubJobs.child("saved_jobs").push();

                            Intent intent2 = new Intent(SingleJobView.this, FeedActivity.class);
                            startActivity(intent2);
                            finish();
                        }

                        dialog.dismiss();
                    }
                });

                builder.setNegativeButton("NO", new DialogInterface.OnClickListener() {

                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                        // Do nothing
                        dialog.dismiss();
                    }
                });

                AlertDialog alert = builder.create();
                alert.show();

            }
        });


    }

    @Override
    protected void onStart() {
        super.onStart();

        workhubJobs.child(job_key).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                FirebaseRecyclerAdapter<Job,SuggesViewHolder> firebaseRecyclerAdapter = new FirebaseRecyclerAdapter<Job, SuggesViewHolder>(
                        Job.class,
                        R.layout.job_row,
                        SuggesViewHolder.class,
                        workhubJobs.orderByChild("jobKeyWord").equalTo((String) dataSnapshot.child("jobKeyWord").getValue()).limitToLast(5)
                ) {
                    @Override
                    protected void populateViewHolder(SuggesViewHolder viewHolder, Job model, int position) {
                        final String job_key = getRef(position).getKey();
                        viewHolder.setJobName(model.getJobName());
                        viewHolder.setJobBudget(model.getJobBudget());
                        viewHolder.setJobLocation(model.getJobLocationName());
                        viewHolder.setJobDate(model.getJobPostedDate());

                        viewHolder.mview.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                Intent intent = new Intent(SingleJobView.this, SingleJobView.class);
                                intent.putExtra("job_id", job_key);
                                startActivity(intent);
                                finish();
                            }
                        });
                    }
                };
                sugges_list.setAdapter(firebaseRecyclerAdapter);
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });



    }

    public static class SuggesViewHolder extends RecyclerView.ViewHolder{

        View mview;
        public SuggesViewHolder(View itemView) {
            super(itemView);

            mview = itemView;
        }

        public void setJobName(String jobname){
            TextView jName = (TextView) mview.findViewById(R.id.jobrowname);
            jName.setText(jobname);
        }

        public void setJobBudget(String jobbudget){
            TextView jBudget = (TextView) mview.findViewById(R.id.jobrowbudget);
            jBudget.setText("Rs."+jobbudget);
        }

        public void setJobLocation(String jobLocation){
            TextView jLocation = (TextView) mview.findViewById(R.id.jobrowlocation);
            jLocation.setText(jobLocation);
        }

        public void setJobDate(String jobDate){
            TextView jDate = (TextView) mview.findViewById(R.id.jobrowdate);
            jDate.setText("Added on "+jobDate);
        }
    }
}

File: app/src/main/java/com/example/himasha/workhub/SingleJobView.kt
package com.example.himasha.workhub

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*

class SingleJobView : AppCompatActivity() {
    private var job_key: String? = null
    private val jobAddedUser: String? = null
    private var jobName: TextView? = null
    private var jobDesc: TextView? = null
    private var jobLocation: TextView? = null
    private var jobBudget: TextView? = null
    private var jobPostedDate: TextView? = null
    private var jobPostedUser: TextView? = null
    private var jobPostedMap: TextView? = null
    private var deletejob: Button? = null
    private var editjob: Button? = null
    private var saveJob: Button? = null
    private var feedbackJob: Button? = null
    private var sugges_list: RecyclerView? = null
    private var locationMapCardview: CardView? = null
    private var editRemoveJobcardview: CardView? = null
    private var postedUserCardview: CardView? = null
    private var workhubUsers: DatabaseReference? = null
    private var workhubJobs: DatabaseReference? = null
    private var workhubSavedJobs: DatabaseReference? = null
    private val workhubsuges: DatabaseReference? = null
    private var auth: FirebaseAuth? = null
    private var builder: AlertDialog.Builder? = null
    private var job_location: String? = null
    private var jobLat: Double? = null
    private var jobLong: Double? = null
    private var posteduserid: String? = null
    private var keyword: String? = null
    private var pref: SharedPreferences? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_single_job_view)
        pref = getSharedPreferences("Users", 0)
        job_key = intent.extras!!.getString("job_id")
        auth = FirebaseAuth.getInstance()
        workhubJobs = FirebaseDatabase.getInstance().reference.child("jobs")
        workhubSavedJobs = FirebaseDatabase.getInstance().reference.child("saved_jobs")
        workhubUsers = FirebaseDatabase.getInstance().reference.child("users")
        builder = AlertDialog.Builder(this)
        jobName = findViewById<View>(R.id.singlejobName) as TextView
        jobDesc = findViewById<View>(R.id.singlejobDesc) as TextView
        jobLocation = findViewById<View>(R.id.singlejobLocation) as TextView
        jobBudget = findViewById<View>(R.id.singlejobBudget) as TextView
        jobPostedDate = findViewById<View>(R.id.singlejobPostedDate) as TextView
        jobPostedUser = findViewById<View>(R.id.singlejobPostedUser) as TextView
        jobPostedMap = findViewById<View>(R.id.singlejobMapName) as TextView
        deletejob = findViewById<View>(R.id.singlejobDeleteBTN) as Button
        saveJob = findViewById<View>(R.id.saveJobButton) as Button
        editjob = findViewById<View>(R.id.singlejobEditBTN) as Button
        feedbackJob = findViewById<View>(R.id.feedbackButton) as Button
        sugges_list = findViewById<View>(R.id.sugges_list) as RecyclerView
        sugges_list!!.setHasFixedSize(true)
        sugges_list!!.layoutManager = LinearLayoutManager(this)
        locationMapCardview = findViewById<View>(R.id.locationMapCard) as CardView
        editRemoveJobcardview = findViewById<View>(R.id.editremoveJobCard) as CardView
        postedUserCardview = findViewById<View>(R.id.postedUserCard) as CardView
        workhubJobs.child(job_key).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot1: DataSnapshot) {
                keyword = dataSnapshot1.child("jobKeyWord").value as String
                val job_name = dataSnapshot1.child("jobName").value as String
                val job_desc = dataSnapshot1.child("jobDesc").value as String
                job_location = dataSnapshot1.child("jobLocationName").value as String
                val job_budget = dataSnapshot1.child("jobBudget").value as String
                val job_postedDate = dataSnapshot1.child("jobPostedDate").value as String
                val job_postedUser = dataSnapshot1.child("jobPostedUserId").value as String
                posteduserid = dataSnapshot1.child("jobPostedUserId").value as String
                val job_postedUserName = dataSnapshot1.child("jobPostedUserName").value as String
                jobLat = dataSnapshot1.child("jobLocationLat").value as Double
                jobLong = dataSnapshot1.child("jobLocationLong").value as Double
                jobName!!.text = job_name
                jobDesc!!.text = job_desc
                jobLocation!!.text = job_location
                jobBudget!!.text = "Rs.$job_budget"
                jobPostedDate!!.text = "Posted on $job_postedDate"
                jobPostedUser!!.text = "Posted by $job_postedUserName"
                jobPostedMap!!.text = "get directions to $job_location"
                if (auth.getCurrentUser()!!.uid == job_postedUser) {
                    editRemoveJobcardview!!.visibility = View.VISIBLE
                } else {
                    editRemoveJobcardview!!.visibility = View.GONE
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
        postedUserCardview!!.setOnClickListener {
            if (posteduserid == auth.getCurrentUser()!!.uid) {
                val intent = Intent(this@SingleJobView, SingleProfileActivity::class.java)
                intent.putExtra("user_id", posteduserid)
                startActivity(intent)
            } else {
                val intent = Intent(this@SingleJobView, SingleProfileActivity::class.java)
                intent.putExtra("user_id", posteduserid)
                startActivity(intent)
            }
        }
        locationMapCardview!!.setOnClickListener {
            val intent = Intent(this@SingleJobView, MapLoadActivity::class.java)
            intent.putExtra("job_lat", jobLat)
            intent.putExtra("job_long", jobLong)
            startActivity(intent)
        }
        editjob!!.setOnClickListener {
            val intent = Intent(this@SingleJobView, EditJobActivity::class.java)
            intent.putExtra("job_id", job_key)
            startActivity(intent)
        }
        deletejob!!.setOnClickListener {
            builder!!.setTitle("Confirm")
            builder!!.setMessage("Are you sure you want to delete this job?")
            builder!!.setPositiveButton("YES") { dialog, which ->
                if (job_key != null) {
                    workhubJobs.child(job_key).removeValue()
                    val intent2 = Intent(this@SingleJobView, FeedActivity::class.java)
                    startActivity(intent2)
                    finish()
                }
                dialog.dismiss()
            }
            builder!!.setNegativeButton("NO") { dialog, which -> // Do nothing
                dialog.dismiss()
            }
            val alert = builder!!.create()
            alert.show()
        }
        saveJob!!.setOnClickListener {
            builder!!.setTitle("Confirm")
            builder!!.setMessage("Are you sure you want to save this job?")
            builder!!.setPositiveButton("YES") { dialog, which ->
                workhubJobs.child(job_key).addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        dataSnapshot.ref.child("isSaved").setValue("1")
                    }

                    override fun onCancelled(databaseError: DatabaseError) {}
                })
                dialog.dismiss()
            }
            builder!!.setNegativeButton("NO") { dialog, which -> // Do nothing
                dialog.dismiss()
            }
            val alert = builder!!.create()
            alert.show()
        }
        feedbackJob!!.setOnClickListener {
            builder!!.setTitle("Confirm")
            builder!!.setMessage("Are you sure you want to save this job?")
            builder!!.setPositiveButton("YES") { dialog, which ->
                if (job_key != null) {
                    workhubJobs.child("saved_jobs").push()
                    val intent2 = Intent(this@SingleJobView, FeedActivity::class.java)
                    startActivity(intent2)
                    finish()
                }
                dialog.dismiss()
            }
            builder!!.setNegativeButton("NO") { dialog, which -> // Do nothing
                dialog.dismiss()
            }
            val alert = builder!!.create()
            alert.show()
        }
    }

    override fun onStart() {
        super.onStart()
        workhubJobs!!.child(job_key).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val firebaseRecyclerAdapter: FirebaseRecyclerAdapter<Job, SuggesViewHolder> =
                    object : FirebaseRecyclerAdapter<Job, SuggesViewHolder>(
                        Job::class.java,
                        R.layout.job_row,
                        SuggesViewHolder::class.java,
                        workhubJobs!!.orderByChild("jobKeyWord")
                            .equalTo(dataSnapshot.child("jobKeyWord").value as String)
                            .limitToLast(5)
                    ) {
                        override fun populateViewHolder(
                            viewHolder: SuggesViewHolder,
                            model: Job,
                            position: Int
                        ) {
                            val job_key = getRef(position).key
                            viewHolder.setJobName(model.jobName)
                            viewHolder.setJobBudget(model.jobBudget)
                            viewHolder.setJobLocation(model.jobLocationName)
                            viewHolder.setJobDate(model.jobPostedDate)
                            viewHolder.mview.setOnClickListener {
                                val intent = Intent(this@SingleJobView, SingleJobView::class.java)
                                intent.putExtra("job_id", job_key)
                                startActivity(intent)
                                finish()
                            }
                        }
                    }
                sugges_list!!.adapter = firebaseRecyclerAdapter
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    class SuggesViewHolder(var mview: View) : RecyclerView.ViewHolder(
        mview
    ) {
        fun setJobName(jobname: String?) {
            val jName = mview.findViewById<View>(R.id.jobrowname) as TextView
            jName.text = jobname
        }

        fun setJobBudget(jobbudget: String?) {
            val jBudget = mview.findViewById<View>(R.id.jobrowbudget) as TextView
            jBudget.text = "Rs.$jobbudget"
        }

        fun setJobLocation(jobLocation: String?) {
            val jLocation = mview.findViewById<View>(R.id.jobrowlocation) as TextView
            jLocation.text = jobLocation
        }

        fun setJobDate(jobDate: String?) {
            val jDate = mview.findViewById<View>(R.id.jobrowdate) as TextView
            jDate.text = "Added on $jobDate"
        }
    }
}